File: leetcode/51.n-queens/cpp/solvenqueens.cpp
#include <vector>
#include <string>

class Solution
{
public:
    std::vector<std::vector<std::string>> solveNQueens(int n)
    {
        board_ = std::vector<std::vector<char>>(
            n, std::vector<char>(n, '.'));
        backtrack(n, 0);
        return res_;
    }

private:
    std::vector<std::vector<std::string>> res_{};
    std::vector<std::vector<char>> board_{};

    bool isValid(int n, int row, int col)
    {
        for (int i = 0; i < row; i++)
        {
            if (board_[i][col] == 'Q')
                return false;
        }
        for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--)
        {
            if (board_[i][j] == 'Q')
                return false;
        }
        for (int i = row - 1, j = col + 1; i >= 0 && j < n; i--, j++)
        {
            if (board_[i][j] == 'Q')
                return false;
        }
        return true;
    }

    std::vector<std::string> joinBoard(int n)
    {
        std::vector<std::string> res(n);

        for (int i = 0; i < n; i++)
            res[i] = std::string(board_[i].begin(), board_[i].end());

        return res;
    }

    void backtrack(int n, int start)
    {
        if (start == n)
        {
            res_.push_back(joinBoard(n));
            return;
        }

        for (int i = 0; i < n; i++)
        {
            if (!isValid(n, start, i))
                continue;

            board_[start][i] = 'Q';
            backtrack(n, start + 1);
            board_[start][i] = '.';
        }
    }
};

File: leetcode/51.n-queens/ts/solvenqueens.ts
const solveNQueens = (n: number): string[][] => {
    const res: string[][] = []
    const board: string[][] = new Array(n).fill(0).map(_ => new Array(n).fill('.'))

    const backtrack = (start: number) => {
        if (start === n) {
            res.push(joinBoard(board))
        }

        for (let i = 0; i < n; i++) {
            if (!isValid(board, start, i)) continue

            board[start][i] = "Q"
            backtrack(start + 1)
            board[start][i] = "."
        }
    }

    backtrack(0)
    return res;
}

const isValid = (board: string[][], row: number, col: number): boolean => {
    const n: number = board.length;
    if (col < 0 || col >= n || row < 0 || row >= n) return false;
    for (let row of board)
        if (row[col] === 'Q') return false;

    let x: number = col,
        y: number = row;
    while (y >= 0 && x < n)
        if (board[y--][x++] === 'Q') return false;

    x = col;
    y = row;
    while (x >= 0 && y >= 0)
        if (board[y--][x--] === 'Q') return false;

    return true;
}

const joinBoard = (board: string[][]): string[] => {
    const res: string[] = []
    board.forEach(row => res.push(row.join('')))
    return res
}

export default solveNQueens
